tools/ja-rule/OpenLightingDevice.cpp

```diff
@@ -0,0 +1,324 @@
+               << m_out_transfer->length << " bytes";
+    }
+  }
+  {
+    MutexLocker locker(&m_out_mutex);
+    m_out_in_progress = false;
+  }
+}
+
+void OpenLightingDevice::_InTransferComplete() {
+  TimeStamp now;
+  Clock clock;
+  clock.CurrentTime(&now);
+  OLA_INFO << "Command completed in " << (now - m_out_sent_time)
+           << ", status is " << libusb_error_name(m_in_transfer->status);
+
+  if (m_in_transfer->status == LIBUSB_TRANSFER_COMPLETED) {
+    // Ownership of the buffer is transferred to the HandleData method,
+    // running on the SS thread.
+    m_ss->Execute(
+        NewSingleCallback(
+          this, &OpenLightingDevice::HandleData,
+          reinterpret_cast<const uint8_t*>(m_in_transfer->buffer),
+          static_cast<unsigned int>(m_in_transfer->actual_length)));
+  } else {
+    delete[] m_in_transfer->buffer;
+  }
+
+  {
+    MutexLocker locker(&m_out_mutex);
+    m_in_in_progress = false;
+  }
+}
+
+bool OpenLightingDevice::SubmitInTransfer() {
+  MutexLocker locker(&m_in_mutex);
+  if (m_in_in_progress) {
+    OLA_WARN << "Reading already pending";
+    return true;
+  }
+
+
+  uint8_t* rx_buffer = new uint8_t[IN_BUFFER_SIZE];
+  libusb_fill_bulk_transfer(m_in_transfer, m_handle, kInEndpoint,
+                            rx_buffer, IN_BUFFER_SIZE,
+                            InTransferCompleteHandler,
+                            static_cast<void*>(this),
+                            kTimeout);
+
+  Clock clock;
+  clock.CurrentTime(&m_send_in_time);
+  int r = libusb_submit_transfer(m_in_transfer);
+  if (r) {
+    OLA_WARN << "Failed to submit input transfer: " << libusb_error_name(r);
+    return false;
+  }
+
+  m_in_in_progress = true;
+  return true;
+}
+
+void OpenLightingDevice::HandleData(const uint8_t* data, unsigned int size) {
+  class ArrayDeleter {
+   public:
+    explicit ArrayDeleter(const uint8_t* data) : m_data(data) {}
+    ~ArrayDeleter() { delete[] m_data; }
+   private:
+    const uint8_t* m_data;
+  };
+
+  ArrayDeleter deleter(data);
+
+  // Right now we assume that the device only sends a single message at a time.
+  // If this ever change from a message model to more of a stream model we'll
+  // need to fix this.
+  if (!m_message_handler) {
+    return;
+  }
+
+  if (size < MIN_RESPONSE_SIZE) {
+    OLA_WARN << "Response was too small, " << size << " bytes, min was "
+             << MIN_RESPONSE_SIZE;
+    return;
+  }
+
+  if (data[0] != SOF_IDENTIFIER) {
+    OLA_WARN << "SOF mismatch, was 0x" << ola::strings::ToHex(data[0]);
+    return;
+  }
+
+  uint16_t command = data[1] + (data[2] << 8);
+  uint16_t payload_size = data[3] + (data[4] << 8);
+  uint8_t return_code = data[5];
+  uint8_t flags = data[6];
+
+  if (payload_size + MIN_RESPONSE_SIZE > size) {
+    OLA_WARN << "Message size of " << (payload_size + MIN_RESPONSE_SIZE)
+             << " is greater than rx size of " << size;
+    return;
+  }
+
+  ola::strings::FormatData(&std::cout, data, size);
```

Comment:
Yeah, it's just for me to debug right now. I'll remove it later - I added a TODO.